src/core/geometry/qgsgeometry.cpp

```diff
@@ -4530,3 +4530,67 @@ bool QgsGeometry::Error::hasWhere() const
 {
   return mHasLocation;
 }
+
+QgsGeometry QgsGeometry::chamfer( int vertexIndex, double distance1, double distance2 ) const
+{
+  const QgsCurve *curve = qgsgeometry_cast<const QgsCurve *>( d->geometry.get() );
```

Comment:
If you want to be more tolerant and also accept multicurves with a single geometry:
```suggestion
  if ( isNull() )
    return QgsGeometry();
  const QgsCurve *curve = qgsgeometry_cast<const QgsCurve *>( d->geometry->simplifiedTypeRef() );
```

elsewhere too